jgrapht-core/src/main/java/org/jgrapht/GraphMetrics.java

```diff
@@ -207,6 +209,123 @@
                 && (graph.getType().isDirected() || graph.getType().isAllowingMultipleEdges());
         return girth;
     }
+
+    /**
+     * An $O(|V|^3)$ (assuming vertexSubset provides constant time indexing) naive implementation
+     * for counting non-trivial triangles in an undirected graph induced by the subset of vertices.
+     *
+     * @param graph the input graph
+     * @param vertexSubset the vertex subset
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     * @return the number of triangles in the graph induced by vertexSubset
+     */
+    private static <V, E> long naiveCountTriangles(Graph<V, E> graph, List<V> vertexSubset){
+        long total = 0;
+
+        for (int i = 0; i < vertexSubset.size(); i++) {
+            for (int j = i + 1; j < vertexSubset.size(); j++) {
+                for (int k = j + 1; k < vertexSubset.size(); k++) {
+                    V u = vertexSubset.get(i);
+                    V v = vertexSubset.get(j);
+                    V w = vertexSubset.get(k);
+
+                    if (graph.containsEdge(u, v) && graph.containsEdge(v, w) && graph.containsEdge(w, u)){
+                        total++;
+                    }
+                }
+            }
+        }
+
+        return total;
+    }
+
+    /**
+     * An $O(|E|^{3/2})$ algorithm for counting the number of non-trivial triangles in an undirected graph.
+     * A non-trivial triangle is formed by tree distinct vertices all connected to each others.
+     *
+     * <p>
+     * For more details of this algorithm see Ullman, Jeffrey: "Mining of Massive Datasets", Cambridge University Press,
+     * Chapter 10
+     *
+     * @param graph the input graph
+     * @param <V> the graph vertex type
+     * @param <E> the graph edge type
+     * @return the number of triangles in the graph
+     * @throws NullPointerException if {@code graph} is {@code null}
+     * @throws IllegalArgumentException if {@code graph} is not undirected
+     */
+    public static <V, E> long getNumberOfTriangles(Graph<V, E> graph){
+        GraphTests.requireUndirected(graph);
+
+        final int sqrtV = (int)Math.sqrt(graph.vertexSet().size());
+
+        List<V> vertexList = new ArrayList<>(graph.vertexSet());
+
+        /*
+            The book suggest the following comparator: "Compare vertices based on their degree.
+            If equal compare them of their actual value, since they are all integers".
+         */
+
+        // Fix vertex order for unique comparison of vertices
+        Map<V, Integer> vertexOrder = new HashMap<>(graph.vertexSet().size());
+        int k = 0;
+        for (V v : graph.vertexSet()) {
+            vertexOrder.put(v, k++);
+        }
+
+        Comparator<V> comparator = Comparator.comparingInt(graph::degreeOf)
+                .thenComparingInt(System::identityHashCode)
```

Comment:
I think identityHashCode is superfluous now since it was necessary to introduce vertexOrder?

Reply:
hashCode is used to introduce vertexOrder and not identityHashCode. I compare on identityHashCode to save calls to Map.get. I only need to use vertexOrder if both the degrees and the identity hash-codes match (which is much more unlike than having just the degrees to match).

Reply:
Got it, make sense.